torch/csrc/autograd/python_saved_variable_hooks.cpp

```diff
@@ -44,4 +44,37 @@ namespace torch { namespace autograd {
       Py_XDECREF(data_);
     }
   }
+
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+  PyObject* PyDefaultSavedVariableHooks::pack_hook_ = nullptr;
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+  PyObject* PyDefaultSavedVariableHooks::unpack_hook_ = nullptr;
+
+  void PyDefaultSavedVariableHooks::set_hooks(py::function &pack_hook, py::function &unpack_hook) {
+    TORCH_CHECK(!pack_hook_ && !unpack_hook_,
+        "Setting default hooks but they have already been set. "
+        "Hint: only one pair of hooks is allowed at a time.");
+    pack_hook_ = pack_hook.release().ptr();
+    unpack_hook_ = unpack_hook.release().ptr();
+  }
+
+  void PyDefaultSavedVariableHooks::reset_hooks() {
+    if (Py_IsInitialized()) {
+      py::gil_scoped_acquire gil;
+      Py_XDECREF(pack_hook_);
+      Py_XDECREF(unpack_hook_);
+    }
+    pack_hook_ = nullptr;
+    unpack_hook_ = nullptr;
+  }
+
+  std::unique_ptr<SavedVariableHooks> PyDefaultSavedVariableHooks::get_hooks() {
+    if (!pack_hook_ || !unpack_hook_) {
+      return nullptr;
+    }
+    py::function pack_hook = py::reinterpret_borrow<py::function>(pack_hook_);
+    py::function unpack_hook = py::reinterpret_borrow<py::function>(unpack_hook_);
+    return std::make_unique<PySavedVariableHooks>(pack_hook, unpack_hook);
```

Comment:
Note that this constructor actually steals the refererence from these two functions. But if you borrow above, then you actually stole the reference from the fields directly.
I feel like there is a missing incref here.

Reply:
These namings :D
borrow actually incref and steal does not.

But isn't the `py::function` a managed type that will get decref automatically? Should you use `pack_hook.release()`?

Reply:
Actually, the `py::function` gets `Py_XDECREF` when it goes out of scope (https://github.com/pybind/pybind11/blob/6d4409466b6c7e6fa53e7b93cb05d78647c26313/include/pybind11/pytypes.h#L199) so we're fine.

If we wanted to use `pack_hook.release()` here, we could do so by adding a new constructor `PySavedVariableHooks:: PySavedVariableHooks(PyObject* pack_hook, PyObject* unpack_hook)` which takes `py::handle` / `PyObject*` (they're the same) as argument.

Reply:
Ho right these arguments are py::function so there is an extra incref when you call the functions...
Mixing pybind and PyObject is tricky!